onchain/rollups/test/foundry/consensus/quorum/QuorumFactory.t.sol

```diff
@@ -0,0 +1,137 @@
+import {Quorum} from "contracts/consensus/quorum/Quorum.sol";
+import {IHistory} from "contracts/history/IHistory.sol";
+import {Vm} from "forge-std/Vm.sol";
+
+contract QuorumFactoryTest is Test {
+    QuorumFactory factory;
+
+    event QuorumCreated(Quorum quorum, IHistory history);
+    struct QuorumCreatedEventData {
+        Quorum quorum;
+        IHistory history;
+    }
+
+    function setUp() public {
+        factory = new QuorumFactory();
+    }
+
+    function testNewQuorum(uint256 _seed, IHistory _history) public {
+        (
+            address[] memory validators,
+            uint256[] memory shares
+        ) = generateAddrAndShares(_seed);
+
+        vm.recordLogs();
+
+        Quorum quorum = factory.newQuorum(validators, shares, _history);
+
+        testNewQuorumAux(validators, shares, _history, quorum);
+    }
+
+    function generateAddrAndShares(
+        uint256 _seed
+    )
+        internal
+        view
+        returns (address[] memory validators_, uint256[] memory shares_)
+    {
+        // let there be no more than 1000 validators so the test wont run forever
+        uint256 numOfValidators = bound(_seed, 1, 1000);
+
+        validators_ = new address[](numOfValidators);
+        shares_ = new uint256[](numOfValidators);
+        for (uint256 i; i < numOfValidators; ++i) {
+            validators_[i] = address(uint160(i + 1));
+            shares_[i] = i + 1;
+        }
+    }
+
+    function testNewQuorumAux(
+        address[] memory _validators,
+        uint256[] memory _shares,
+        IHistory _history,
+        Quorum _quorum
+    ) internal {
+        Vm.Log[] memory entries = vm.getRecordedLogs();
+
+        uint256 numQuorumCreated;
+        for (uint256 i; i < entries.length; ++i) {
+            Vm.Log memory entry = entries[i];
+
+            if (
+                entry.emitter == address(factory) &&
+                entry.topics[0] == QuorumCreated.selector
+            ) {
+                ++numQuorumCreated;
+                QuorumCreatedEventData memory eventData = abi.decode(
+                    entry.data,
+                    (QuorumCreatedEventData)
+                );
+                assertEq(address(eventData.quorum), address(_quorum));
+                assertEq(address(eventData.history), address(_history));
+            }
+        }
+        assertEq(numQuorumCreated, 1);
+
+        assertEq(_quorum.numOfValidators(), _validators.length);
+        assertEq(address(_quorum.getHistory()), address(_history));
+        address[] memory returnedValidators = _quorum.validators();
+        assertEq(returnedValidators.length, _validators.length);
+        uint256 totalShares;
+        for (uint256 i; i < _validators.length; ++i) {
+            assertEq(returnedValidators[i], _validators[i]);
+            assertEq(_quorum.shares(returnedValidators[i]), _shares[i]);
+            totalShares += _shares[i];
+        }
+        assertEq(_quorum.totalShares(), totalShares);
```

Comment:
I've pushed a change to `feature/quorum` which lets you calculate the sum of arrays more easily. This can simplify your code.

https://github.com/cartesi/rollups-contracts/blob/a8454c6ecd38f2cb7a84def3625bd5d369579f18/onchain/rollups/test/foundry/consensus/quorum/Quorum.t.sol#L122